content types - ready to review (#37)
SenseNet/docs.sensenet.com: content/concepts/content-management/03-content-types.md

metaDescription: "sensenet Content Types"
---

The content repository contains many different types of content. Content vary in structure and even in function. Different types of content contain different fields and may also implement different business logic. The fields and business logic of a content is defined by its type - the content type.

# Content Types

A content type is a special content in the content repository. Content types define the structure and functions of contents. It is defined by a (xml like) configuration file called ``content type definition`` or ``CTD``.

For example a User has a name, e-mail address, etc. - these fields of the user ``content type`` are defined by its ``content type definition``.

> Additional logic, like a specific task when a user is saved can be implemented in the ``content handler``.

# Default content types

> In sensenet we provide a default set of predefined content types. These types can be used in any solution built with sensenet but additional (custom) content types can be configured as well according to business needs.

Reviewer: nyelvezet: "we provide" helyett inkább passzív, de nem létkérdés

> In sensenet we provide a default set of predefined content types. These types can be used in any solution built with sensenet but additional (custom) content types can be configured as well according to business needs.

Author: ```suggestion
A default set of predefined content types is provided in sensenet. These types can be used in any solution built with sensenet but additional (custom) content types can be configured as well according to business needs.
```

some example from the default set:
- user
- folder
- file
- workspace
- image
- ...

# Custom content type

> Since many feature are tied to content types (AllowedChildType, content template) it is strongly suggested to create custom content types according to your solution's architecture. Default content types can be used in your solution but cannot be modified (in SNaaS). You can create custom content type by inheritng from ``GenericContent`` or any other content type in the tree.

Reviewer 2: ```suggestion
Since many feature are tied to content types (AllowedChildType, content template) it is strongly suggested to create custom content types according to your solution's architecture. Default content types can be used in your solution but cannot be modified or deleted (in SNaaS). You can create custom content type by inheriting from ``GenericContent`` or any other content type in the tree.
```

# Content type hierarchy

> Content types are organized into hierarchy according to inheritance. Any content type may inherit from another one. The topmost content type in the inheritance hierarchy is the ``GenericContent`` (with handler ``SenseNet.ContentRepository.GenericContent``), every content type must inherit from this, or any of its descendant. When a child content type is inherited from a parent content type it means that the child content type contains all the fields of the parent, even if they are not defined in the child CTD - and also they share common implemented logic.

Reviewer 2: ```suggestion
Content types are organized into a hierarchy according to inheritance. Any content type may inherit from another one. The topmost content type in the inheritance hierarchy is the ``GenericContent`` (with handler ``SenseNet.ContentRepository.GenericContent``), every content type must inherit from this, or any of its descendant. When a child content type is inherited from a parent content type it means that the child content type contains all the fields of the parent, even if they are not defined in the child CTD - and also they share common implemented logic.
```

> Multiple inheritance is not allowed so content types are arranged in a simple tree.

## Field inheritance

> A content type inherits its fields from its parent content type (defined by the ``parentType`` attribute). This means that only additional fields have to be defined in the type’s CTD. The inherited fields apply to the content type as defined on the parent type, but may also be overridden. The following apply to field inheritance:

Reviewer 2: ```suggestion
A content type inherits its fields from its parent (and upper ancestor) content type (defined by the ``parentType`` attribute). This means that only additional fields have to be defined in the type’s CTD. The inherited fields apply to the content type as defined on the parent type, but may also be overridden. The following apply to field inheritance:
```

- fields of all ancestors are inherited: ie. fields of parent type of the direct parent are also available in the current type
- all fields of the parent type are inherited: deleting a field that has been defined on an ancestor type is not possible
- a field is inherited from parent when it is not defined in the current type’s CTD
- if a field is defined in a CTD it overrides parent’s field of the same name
- if a field is defined in a CTD with empty markup the parent’s field of the same name is overridden with empty data
- when inheriting a field the first order elements of the configuration element are inherited (these are defined by the field definition)

> > Every detail of a field (e.g: name, visibility, compulsory) can be ovverriden on child level allowing you to break inheritance field by field.

Reviewer 2: ```suggestion
> Every detail of a field (e.g: name, visibility, compulsory) can be overridden on child level allowing you to break inheritance field by field.
```

```xml
<?xml version="1.0" encoding="utf-8"?>
<ContentType name="" handler="" parentType="" xmlns="http://schemas.sensenet.com/SenseNet/ContentRepository/ContentTypeDefinition">
  <DisplayName></DisplayName>
  <Description></Description>
  <Icon></Icon>
  <Preview></Preview>
  <AllowIndexing></AllowIndexing>
  <AppInfo></AppInfo>
  <AllowIncrementalNaming></AllowIncrementalNaming>
  <AllowedChildTypes></AllowedChildTypes>
  <Fields>
    <Field name="" type="">
      <DisplayName></DisplayName>
      <Description></Description>
      <AppInfo></AppInfo>
      <Bind property="" />
      <Indexing>
         <Mode></Mode>
         <Store></Store>
         <TermVector></TermVector>
         <Analyzer></Analyzer>
         <IndexHandler></IndexHandler>
      </Indexing>
      <Configuration>
        <ReadOnly></ReadOnly>
        <Compulsory></Compulsory>
        <DefaultValue></DefaultValue>
        <VisibleBrowse></VisibleBrowse>
        <VisibleEdit></VisibleEdit>
        <VisibleNew></VisibleNew>
        <FieldIndex></FieldIndex>
        <ControlHint></ControlHint>
      </Configuration>
    </Field>
  </Fields>
</ContentType>
```

> The content type definition xml of a content type can be edited on the admin ui or through OData REST API ([see How to create a Content Type for details](https://community.sensenet.com/docs/tutorials/how-to-create-a-content-type/)).

Reviewer 2: ```suggestion
The content type definition xml of a content type can be edited on the admin ui or through OData REST API ([see How to create a Content Type for details](/tutorials/content-types/how-to-create-a-content-type/)).
```

# Usage (advantages)

> Sensenet content type approach enables you to define custom content types in a system according to your needs. This allows great flexibility on content level.  You can freely use predefined content types available in the system or create your own by inheriting from ``GenericContent`` or any other existing content type. It is recommended to define the required custom content types right at the beginning of your project and build your solution afterwards.

Reviewer 2: ```suggestion
Sensenet content type approach enables you to define custom content types in a system according to your needs. This allows great flexibility on content level. You can freely use predefined content types available in the system or create your own by inheriting from ``GenericContent`` or any other existing content type. It is recommended to define the required custom content types right at the beginning of your project and build your solution afterwards.
```
